src/qibo/backends/__init__.py

```diff
@@ -71,26 +83,12 @@ def set_backend(backend="custom"):
     Args:
         backend (str): A backend from the above options.
     """
-    if backend not in AVAILABLE_BACKENDS:
-        available = ", ".join(AVAILABLE_BACKENDS)
-        raise_error(ValueError, "Unknown backend {}. Please select one of the "
-                                "available backends: {}."
-                                "".format(backend, available))
-    if not config.ALLOW_SWITCHERS and backend != K.gates:
+    bk = _construct_backend(backend)
+    if not config.ALLOW_SWITCHERS and backend != K.name:
         warnings.warn("Backend should not be changed after allocating gates.",
                       category=RuntimeWarning)
-
-    gate_backend = backend.split("_")
-    if len(gate_backend) == 1:
-        calc_backend, gate_backend = _BACKEND_NAME, gate_backend[0]
-    elif len(gate_backend) == 2:
-        calc_backend, gate_backend = gate_backend
-    if gate_backend == "custom":
-        calc_backend = "tensorflow"
-    bk = _construct_backend(calc_backend)
     K.assign(bk)
```

Comment:
If we replace this line with `K = bk` the bug in #374 seems to be fixed.

This sounds like a problem with the __init__ global allocation and the `_CONSTRUCTED_BACKENDS`. 
One possibility to avoid this issue consists in always reuse the backends from `_CONSTRUCTED_BACKENDS`.

Reply:
I just pushed some updates in the backend switching mechanism that fixes the device issue and avoids the `.assign` function. I think the idea is close to one of the solutions we were discussing earlier today. I still need to polish several things, fix the tests and your above comments but let me know if you have any comments on this implementation.

Reply:
Thanks, looks pretty good, and indeed it is quite close to my draft in #374. I think we can keep your approach (without the new method), given that you already expose all required features!